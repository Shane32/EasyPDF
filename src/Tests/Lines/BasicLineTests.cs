using System.Drawing;
using Font = Shane32.EasyPDF.Font;

namespace Tests.Lines;

public class BasicLineTests
{
    private readonly PDFWriter _writer = new PDFWriter();

    public BasicLineTests()
    {
        _writer.ScaleMode = ScaleModes.Inches;
        _writer.NewPage(PageKind.Letter, false, 1f, 1f);
        _writer.PrepForTests();
    }

    [Fact]
    public void FillStyles()
    {
        _writer.FillColor.ShouldBe(Color.Black);

        DrawGeometricFigure(0.5f, false, false, false, false, 0.1f, LineJoinStyle.Miter, 0, LineCapStyle.None);

        _writer.MoveTo(3.75f, 0);
        DrawGeometricFigure(0.5f, true, true, false, false, 0.1f, LineJoinStyle.Miter, 0, LineCapStyle.None);

        _writer.MoveTo(0, 2);
        DrawGeometricFigure(0.5f, true, false, true, false, 0.1f, LineJoinStyle.Miter, 0, LineCapStyle.None);

        _writer.MoveTo(3.75f, 2);
        DrawGeometricFigure(0.5f, true, false, true, true, 0.1f, LineJoinStyle.Miter, 0, LineCapStyle.None);

        _writer.MoveTo(0, 4);
        DrawGeometricFigure(0.5f, true, true, true, true, 0.1f, LineJoinStyle.Miter, 0, LineCapStyle.None);

        _writer.MoveTo(3.75f, 4);
        DrawGeometricFigure(0.5f, true, true, true, false, 0.1f, LineJoinStyle.Miter, 0, LineCapStyle.None);

        _writer.MoveTo(0, 6);
        DrawGeometricFigure(0.5f, true, false, false, false, 0.1f, LineJoinStyle.Miter, 0, LineCapStyle.None);

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }


    [Fact]
    public void LineStyles()
    {
        _writer.ForeColor.ShouldBe(Color.Black);
        _writer.LineStyle.JoinStyle.ShouldBe(LineJoinStyle.Miter);
        _writer.LineStyle.CapStyle.ShouldBe(LineCapStyle.None);
        _writer.LineStyle.DashStyle.ShouldBe(LineDashStyle.Solid);
        _writer.LineStyle.Width.ShouldBe(0.1f, 0.001f);

        _writer.MoveTo(0, 0f);
        DrawGeometricFigure(0.5f, false, true, false, false, 0.1f, LineJoinStyle.Miter, 0, LineCapStyle.None);
        _writer.MoveTo(3.75f, 0f);
        DrawGeometricFigure(0.5f, true, true, false, false, 0.1f, LineJoinStyle.Miter, 0, LineCapStyle.None);

        _writer.MoveTo(0, 2f);
        DrawGeometricFigure(0.5f, false, true, false, false, 0.0625f * 72, LineJoinStyle.Miter, 0, LineCapStyle.None);
        _writer.MoveTo(3.75f, 2f);
        DrawGeometricFigure(0.5f, true, true, false, false, 0.0625f * 72, LineJoinStyle.Miter, 0, LineCapStyle.None);

        _writer.MoveTo(0, 4f);
        DrawGeometricFigure(0.5f, false, true, false, false, 0.0625f * 72, LineJoinStyle.Rounded, 0, LineCapStyle.Round);
        _writer.MoveTo(3.75f, 4f);
        DrawGeometricFigure(0.5f, true, true, false, false, 0.0625f * 72, LineJoinStyle.Rounded, 0, LineCapStyle.Round);

        _writer.MoveTo(0, 6f);
        DrawGeometricFigure(0.5f, false, true, false, false, 0.0625f * 72, LineJoinStyle.Bevel, 0, LineCapStyle.Square);
        _writer.MoveTo(3.75f, 6f);
        DrawGeometricFigure(0.5f, true, true, false, false, 0.0625f * 72, LineJoinStyle.Bevel, 0, LineCapStyle.Square);

        _writer.LineStyle.Width = 0.1f;
        _writer.MoveTo(0, 0f);
        _writer.ForeColor = Color.Blue;
        _writer.LineTo(0, 8f);

        _writer.NewPage(PageKind.Letter, false, 1f, 1f);

        _writer.MoveTo(0, 0f);
        DrawGeometricFigure(0.5f, false, true, false, false, 0.02f * 72, LineJoinStyle.Miter, 0, LineCapStyle.None);
        _writer.MoveTo(3.75f, 0f);
        DrawGeometricFigure(0.5f, false, true, false, false, 0.02f * 72, LineJoinStyle.Miter, 1, LineCapStyle.None);

        _writer.MoveTo(0, 2f);
        DrawGeometricFigure(0.5f, false, true, false, false, 0.02f * 72, LineJoinStyle.Miter, 2, LineCapStyle.None);
        _writer.MoveTo(3.75f, 2f);
        DrawGeometricFigure(0.5f, false, true, false, false, 0.02f * 72, LineJoinStyle.Miter, 3, LineCapStyle.None);

        _writer.MoveTo(0, 4f);
        DrawGeometricFigure(0.5f, false, true, false, false, 0.02f * 72, LineJoinStyle.Miter, 4, LineCapStyle.None);

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    private void DrawGeometricFigure(float scale, bool close, bool border, bool fill, bool eofill, float lineWidth, LineJoinStyle joinStyle, int dashStyle, LineCapStyle capStyle)
    {
        var pos = _writer.Position;
        var len = 0.25f * scale;
        var leg = (float)Math.Sqrt(Math.Pow(len, 2) / 2);

        _writer.ForeColor = Color.Red;
        _writer.LineStyle.Width = lineWidth;
        _writer.LineStyle.JoinStyle = joinStyle;
        if (dashStyle == 0) {
            _writer.LineStyle.DashStyle = LineDashStyle.Solid;
        } else if (dashStyle == 1) {
            _writer.LineStyle.DashStyle = LineDashStyle.Dash;
        } else if (dashStyle == 2) {
            _writer.LineStyle.DashStyle = LineDashStyle.DashDotDot;
        } else if (dashStyle == 3) {
            _writer.LineStyle.DashStyle = LineDashStyle.Dot;
        } else if (dashStyle == 4) {
            _writer.LineStyle.DashStyle = new LineDashStyle(new float[] { 1f, 1f, 2f, 1f, 3f, 5f }, 5f);
        }
        _writer.LineStyle.CapStyle = capStyle;

        _writer
            .LineTo(1f * scale, 0f)
            .CornerTo(.25f * scale, .25f * scale, false)
            .LineTo(0f, 0.5f * scale)
            .BezierTo(0f, 0.25f * scale, 0.25f * scale, 0.25f * scale)
            .LineTo(0.5f * scale, 0)
            .BezierTo(0.25f * scale, 0f, 0.25f * scale + leg, leg)
            .LineTo(0.5f * scale, 0.5f * scale)
            .BezierTo(.25f * scale, .25f * scale, .75f * scale, -0.25f * scale, 1f * scale, 0f)
            .LineTo(0f, 1f * scale)
            .LineTo(-2f * scale, 0f)
            .LineTo(0f, .5f * scale)
            .LineTo(.5f * scale, 0f)
            .LineTo(0f, -1f * scale)
            .LineTo(0.5f * scale, 0f)
            .LineTo(0f, 1.5f * scale)
            .LineTo(-1f * scale, 0f)
            .CornerTo(-0.25f * scale, -0.25f * scale, true);

        if (close) {
            _writer.FinishPolygon(border, fill, eofill);
            _writer.Position.ShouldBe(pos);
        } else {
            _writer.FinishLine();
            _writer.Position.ShouldNotBe(pos);
        }

        _writer.MoveTo(pos);
        _writer.OffsetTo(4.5f * scale, 0).Rectangle(1f * scale, 0.4f * scale, 0, fill, border);
        _writer.OffsetTo(0, 0.7f * scale).Rectangle(1f * scale, 0.5f * scale, 0.2f * scale, fill, border);
        _writer.OffsetTo(0, 0.8f * scale).RectangleDualOffset(1f * scale, 0.7f * scale, 0.15f * scale, 0);
        _writer.OffsetTo(0, 1f * scale).RectangleDualOffset(1f * scale, 0.7f * scale, 0.15f * scale, 0.05f);
    }

    [Fact]
    public void LineDashStylesCompare()
    {
        var style1 = new LineDashStyle(new float[] { 1f, 1f, 2f, 1f, 3f, 5f }, 5f);
        style1.Array.ShouldBe(new float[] { 1f, 1f, 2f, 1f, 3f, 5f });
        style1.Array[0] = 50f;
        style1.Array.ShouldBe(new float[] { 1f, 1f, 2f, 1f, 3f, 5f });
        var style2 = new LineDashStyle(new float[] { 1f, 1f, 2f, 1f, 3f, 5f }, 5f);
        (style1 == style2).ShouldBeTrue();
        style1.Equals(style2).ShouldBeTrue();
        style2.Equals(style1).ShouldBeTrue();
        var style3 = new LineDashStyle(new float[] { 1f, 1f, 2f, 1f, 3f, 5f }, 6f);
        (style1 == style3).ShouldBeFalse();
        style1.Equals(style3).ShouldBeFalse();
        style3.Equals(style1).ShouldBeFalse();
        var style4 = new LineDashStyle(new float[] { 1f, 1f, 2f, 1f, 3f, 5f, 6f }, 5f);
        (style1 == style4).ShouldBeFalse();
        style1.Equals(style4).ShouldBeFalse();
        style4.Equals(style1).ShouldBeFalse();
    }

    [Fact]
    public void InvalidLineDashStyleThrows()
    {
        Should.Throw<ArgumentNullException>(() => new LineDashStyle(null!, 0f));
    }
}
