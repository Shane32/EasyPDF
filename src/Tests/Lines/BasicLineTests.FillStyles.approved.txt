%PDF-1.4
%\xE2\xE3\xCF\xD3
1 0 obj
<</Length 2013/Filter/FlateDecode>>stream
x\x9CuYK\xCF\xDCD\x10\xBC\xFBW\xF8\x1C\x09kfz\xC6\x8F+\x12B\xE2\x06\xE4\x86\xB8\xB0\x01$\xB4A\x02$\xF8\xFBLwU\xB7\xC7\xDE\x8D\x92|[\xB5\xDF<z\xBC\xEE\xAAZ\xE7\xAF\xE9\xEB\x8F\x93\xAC\xF3\xD6\xD6\xF9\xE3\xA7\xE9\x9B\x8F\xD3\xF7S\x9ES\xFF\xF3U\x9E\xB7\xD2\xFF\xA6\xF9\xF1yJ\xF3w\xFD\xDF\x1FSZ\xF2\xFC\xDF\xF4\xD3\xCF\xFD\xF7\x9F\xA6d\xE3~\xF8\x96\xE0\xEF\xDF\xDF\x0F\xCB\xE3\xB0\xCF\xBAY\x9A\x9FSM\xCB\xB1uT\xDB\\\x97$\xFAz\xCC\x8F\xA9\xBF\x94M\x7F\xDD\xE6>\xB0U\xFD\xF9\xEF\xD4\x0B\xE9\xAF\xCFi\xCF\xFA\xBAoK\xFFY\xCBb\xBF\xCB\xA9)X\xD3bCr\xAEF\x0F\xFD\x99\xC5\x06\xB5l\xA4\xE6s\xE0c"=V\xCC\xC3\x84+\xE3Z}\xDF\xED\xDDv\xEE\x88\x02n\x94[\xC7R7Z\xFA\x12\xC7\xBC\xF6\xA1\xD2\x89\x9C\x85\x91<\xA6\x1F\xA7\xBC\x16^\x81\xD2\x962\xFF\xFD\xAB\xBD\xB7-unG\xFFao\xF6\x92\xAA\xFF\xA6-\xFD\x00z\x89\xF3\xA1\xEF\x1Ez\x99\xB3\x9E\xE8P\x9C\x8F}>\xF2\xB2f Y\xEC"t\x98s\x7F\xD7\x86*\x96\xA5\x1D3g\xE5l[\xD8j\xC0\x0Fn\x03\xD6\xE7\xACb+b`\xAF\xD7\x170\x98\xB1G\xC7\xB6\xDD\x13\x105`\xA2\xD6\xE5u\xDB\x91\xB7\xFE^\xEB\xDB\xF5C\xEC\x07\xA0V\xD6o\x16\xB2|\x14\x1D\xBE-);>\xEC 
sZQ\x16\xD9\xBEd=\x8C\xCD\xCE\xE9X\xFA\x01wG\x0F\xDB\x0CX+;lE\x0C\xB2\x8B\xEC\xD3IV\x1E_Y_\xEE\xE9\x10\x95`:*\xF43\xD8\x81\xFC\xE6/\x1Bo\xFF\x84\xFB_2\x1B@2;@\x81\x1EE_\xAD\x07\x14\xF4\x0F_
\xBB@*\xDB@\x9A\xF5\x81\xB4K#\xC8vi\x04\xD9\x87F\xA8ih\x84\x9A/\x8D@\xEA\xB7\xBE\xC8\xF1\x86\xC6\xBD\xCFM\xEF\xDC\xEF~\x16q\xA3q\xF7\xFBjw\x8Ev\x10\x19\xFBA\xE4\xDA\x10\xFFLU\x8A^\xB0u\xB8\xEBk_\xA0\xA0\x15>O\xB5\x1Fi\x07\xE9
\xB2\xD6e#\xABk\x07\xFBR\x81\xFA\xE6\xFA\xE1\x18\\md\x07\xC7\xD2o\x7F\xCE\xA9bst1\x83\x0Fnc\xA4\x8F\xEFe\x162-\x89s\xC5>\x9D\x87Ui[<r_\xCCA5g\xCD\x0F;\xC2k\x9BW\xF9R\x9B\xEBBls\xAD\x96m^Wos=\x0CZ\xCC\x10\xDA\\\xA1\xB79\xB0u)g\xA1{\xB1\x9A\xB79\xB6\xF16\xAF2\xB4\xB9\xD6\xEB\x0B\xC8\xD9\xE6\x8A\xD9\xE6\x06Q\x83x\x9BG\xDDv\xE4\x1Am^[\xB4\xB9^qo\xF3\xBA\x9EmNl\xF6\xB0\x8Em\xEE\xCC\xFA\x94\xB3\xD1\xC1\xBA\xAA\xB7\xB9n\xE6m\xAE\x97>\xDA\x1C\x17\xD9\xA7\xCB\xD8\xE6`\xD6\xE6\x84\xA8D\xCE6\x8F3\\\xDA\xBC\xEBY\xADp9\x05\xEEs\x8A\xBB\xA5\xE5\xBA\xD3\xEBr\x13\xB8\x9D\xAA\x85\xD9]\xDE\x93\xDE\x06\xFAb\x86\xA7\xC0\x1CO\x01h_/\x9EWR\xBD\x98^\xC92tW>\xDA`{\x1C\x1B\xBEWj\x1A\xE9N\xDB>Z\xDD\x9D\x96r\xF1\xBA\x17\xBE]l\xF4\x85\xE6\xD3\xFB\xCA\xBA\x9DZD\xF2\x98~3\xA7\xD0\x8B64\xFBo&\xD2e\xC6\xCF\xCF\xE6
;\xD9\xD3<osjV\xB6\x15m<\xC0\xD5>'\xC3{r\xAF\xDBE\xFB\xD6'\xF6\xA3\x95\x99k\x1A~\xF8~\xC6\x9E\xE6|\xC5\xA9\xD5\xC7\x05V|Xp:\xEC\xF5$F\x09\x9C8T\x8F\x9Fq\xD0c\x7F\xE7\xF2%\xC9\x97l\xBEH\x9FW\x1CF\xAFD\x97+\xDAn\x99p\x0B\xAF/\xAD\x85\xD7\x97\xAE\xDC\xA7\xD7\x97v\x84\xD7\x03\xBB\xD7\x83\x85\xD7s\xE0Zb\x01\x83-\xBC\x1E\xFB=\x89\xDB\xE0\xF6V\\\xD4\xEF~_\xE4\x08\xC3\x07\x0E\xC7\x07\x85\x9B\x97~\xDF\x86\xE7\x97*a\xFA\xA5\xA5\xC1\xF4K+\x83\xE9\x97&4} \x98>p\x98>\x07\xE1\x9As:I\x1AL_\xF7|\x06\xCE\x83\xED\xB3\xCC8\xCB\x8B\xF1S\x13\x92\x8B\x02\xCD\xDFn\xF0\x1C\xB2`\xD0t\xC1\x90\x09\x83\xA1~si\x08\x804h
\xC0M\xAC1\xC0~\xD7\xAE\xEA@\xD3u`\x12\xA0:0
P\x1Dh\xFE\xA1\x0E\xCE]\x0Eh\xD0/\xDC\x05\x81{\xBFpW\x04\xAF\xE5\xCE\xB7k\xBCx\xE5y\x08\x04\xD4\x05\x06\x82S%>\x98\xD3\xBC\xC8\x04\xFC\xD5e\x02>\xEE2A\x87\x075+D\x8F\x02\xAE\x11\x0Cp\x85\x81d\x88\x06\xE8~\xAEI\x99\xE0~\x94\x09:=\x07J\xC8\x04`\xA2S\xBAL\x00\x97!"\xB04\xB9\xCA\x84\x8D\xBB\xCB\x04\xAC\xE9\xADL\xC0n!\x13pv\xC8\x04-_\x89\x85"\xF4'\xE0\x16Y\xC1e\x02x\x1B\xB2\x02\xBA\x9F+R&\xB8\x17e\x82\x96\xCF\x81\x122\x01\xD8"+\xB8L\x00\xB7!-Xq2\xCA\x84z\xAD\xCB\x84Z\xBB\xCB\x04=\x1F\x14i\x80\x9D\xE9D"4\xB8L8+Ch\x80\x02\xD8\xCA\x94\x09\xDB\x912A\xD7\xE7 \x19d\xC2I\x1AB\x03d\xC2q\x1Eb\x03\xCB\xACoe"\xF5\xA0\xB8#8(\xF0\xE0\xA0X\xBF\x13wmCp(\xFDM\x0B\x0E\xAA\x0C\xF8\x9E\xDC\xA5A\xF3c\x7F\xC17\xE5R\xF9U\xB9\x83}\xEC\x9C\x08\x0ER\xFD\xCB*\x82\x83+\x08\x82\x83\x87r\x04\x05\x8E\x8D\xE0@5q7\xBFQ\xAA\x8B'\x85\x1B\x95u\xBD\x04\x85;\xAF\xE5\x92B\xEE\xB4\xC7\xCF\x08\x0E\x97/1$\x8F\xE9\x97\x0F\xB0\x9D}\x1F%\xE1k:\xB9d\x19\x92\x03X$\x07P\xEB\xB1\xBCzr\x90^\x81'\x07\\W\xA0mH\x0E"9\x92\x03\xB0'\x07\xB0H\x0E\x1C\xB8\x96X`\xC5\xA7\x05\xDF\xC4^\xF0M/\x81\x13Q\xDAp\x86\x87\x9D\xA93|%\xBC'\x07\xA9\xDB\x97\x92\x83l{$\x07\xC5\x91\x1C\x94\xD8\xE1\xB6\xC3\x93\x83\xEC9\x92\x83\x1CG$\x87\x9A\xF2\x90\x1Cj\x92H\x0E\xC0\x9E\x1C\xC0"9p`/\xDB\x17\xD0\x13\xD8\xBA\xBC\x02{\x8E\xE4\xE0e`\xAA\x15\x17\xF5{r\x90]"9\x00Gr\x00E$\x90\xBD\x9D\xC9A\xF6-\x92\x83\xF4\x9A\xCF\xE4 \xFD8gr\x90\x9E\xF8\x91\x1C\x80\x90\x1C\x80#9p\x10\xAE9\xA7\x93\xD4!9\xE8\x9E\xCF\xC0mH\x0E,3\xCE\xF2\x92\x1C(
\xC9U\x81\xC9\xC1n\xF0\x1C\xBA\x00\xB819@\x19\x0C\x95\x1A\xFF\xF8\xFC\xC0nb{\x80\xA0\xEF\xB7\xAB<\xD0\x9DC\x1E\xD8\xDC\xFE\xA0!\x8D\xF2\xC0\x9E\x0Byp\xEEz\xC0aw\x1E\x8Ap]4xH\x82\xD7r\xE3\xA1\x09\xBE\xDE\x9DC$\x98\x1C\xBC(\xB9\xC9\x04\xBF\x99\xDFT\x02\xA6\xEB*\x01\x93w\x95\xA0\xFF\x83\xDA\x13\x03\xB4(\x1E(\xA4\xF3\x89B\xA9\xFEH\x01M\xEE\x13\xAD\xF9\xB9&U\x82\xFBQ%\xE8\xFF\x1C(\xA1\x12\x805\x9E,\xA4\x08\x0E^\x82\x0C*1\x9E\xE1agz\xA3\x12p\xA7\xB7*\x01\xE3\x85J\xC0\xE6\xA1\x12\xF4\x7F%v8\xB4\xA7A\xAA\x04\x1E\xA3\x1C\x11\x1C\xD8\xE4\x9C\x88\xE6\xC7\x8A\xAE\x12\xD8\xCBU\x82\xEE\xCF\x81\x12*\x81G,G\x04\x07W\x09\xC3,CB%\xCE\xFA=8\xB8J\xA8\xBD\xBBJ\xD0\xF7A\x91\x08\xD8\x98N\xB6\x08\x0E\xAE\x12\xCE\xD6!8@\x00le\xAA\x84\xEDH\x95\xA0\xEFs\x90\x0C*\xE1\xA4\x0E\xC1\x01*\xE1\xB8\xC1\x81e\xD6\xB7*\x91\xEC\x02YpP\xE0\xC1\xC1.`\xEB/+\x83\x83N\xB7\xE0\xA0\xD5Zp\xD0\x0F\xA9!\x07ZpP`\xC1A\x01\xFD{\xAB\x97\xE0P\x0F\xFF\xA6\x8F\xE0\xD0R\x1E\x82C\xDDe\x08\x0E\x1C\x1B\xC1\xA1\x95}t\xF3;\xAD\xEB\x98\x14\xEE4\xFB\xE3}\xD4\xF1\xC2\xD7K
\xB9\xD3\x9EE#8\xB46<&%yL\x7F\x9A\xE9\xD4}}g\xA7\xF5\xC8_\xB2\xD3VJ\xD8\xA9\xE2\xB0S%\xEA\x95\xAD\x88\xDBi+-\xEC\xB4U\x09;m\xB5v\xDA\xEA\x16v
\xECv
\x16v\xCA\x81\xF6\x9F\x04\xCD\xED\x14\xEB\xC2N\xB1\xDF\x93X\x06;\xB5\xE2\xA2~\xB7\xD3V\xB6\xB0S\xE0\xB0SP\xF8d+\xC7i\xA7Mr\xD8i\x93}\xB0\xD3\xD6\xEF\xC3\xD3N[\xBF\x07`\xA7@\xB0S\xE0\xB0S\x0E\xC2\x7Fqp:\x88\xAD\xECv\xAA{\xBA\x9Dz1\\\x00e\xC6Y\xF4d\xFF\x03\x07\xE0!\xEA
endstream
endobj
3 0 obj
<</Type/Page/MediaBox[0 0 612 792]/Resources<</ProcSet [/PDF /Text /ImageB /ImageC /ImageI]>>/Contents 1 0 R/Parent 2 0 R>>
endobj
2 0 obj
<</Type/Pages/Count 1/Kids[3 0 R]/ITXT(3.3.2.0)>>
endobj
4 0 obj
<</Type/Catalog/Pages 2 0 R>>
endobj
5 0 obj
<<>>
endobj
xref
0 6
0000000000 65535 f 
0000000015 00000 n 
0000002235 00000 n 
0000002096 00000 n 
0000002300 00000 n 
0000002345 00000 n 
trailer
<</Size 6/Root 4 0 R/Info 5 0 